static/app/components/arithmeticBuilder/action.tsx

```diff
@@ -50,84 +50,77 @@ function isArithmeticBuilderReplaceAction(
 }
 
 export interface ArithmeticBuilderState {
+  expression: string;
   focusOverride: FocusOverride | null;
-  query: string;
+  tokens: Token[];
+  validated: 'valid' | 'invalid';
 }
 
 interface UseArithmeticBuilderActionOptions {
-  initialQuery: string;
+  initialExpression: string;
+  updateExpression?: (expression: Expression) => void;
 }
 
 export function useArithmeticBuilderAction({
-  initialQuery,
-}: UseArithmeticBuilderActionOptions) {
-  const reducer: Reducer<ArithmeticBuilderState, ArithmeticBuilderAction> = useCallback(
-    (state, action): ArithmeticBuilderState => {
-      if (isArithmeticBuilderUpdateResetFocusOverrideAction(action)) {
-        return resetFocusOverride(state);
-      }
-
-      if (isArithmeticBuilderDeleteAction(action)) {
-        return deleteToken(state, action);
-      }
+  initialExpression,
+  updateExpression,
+}: UseArithmeticBuilderActionOptions): {
+  dispatch: (action: ArithmeticBuilderAction) => void;
+  state: {
+    expression: Expression;
+    focusOverride: FocusOverride | null;
+  };
+} {
+  const [expression, setExpression] = useState(() => new Expression(initialExpression));
+  const [focusOverride, setFocusOverride] = useState<FocusOverride | null>(null);
 
-      if (isArithmeticBuilderReplaceAction(action)) {
-        return replaceToken(state, action);
+  const dispatch = useCallback(
+    (action: ArithmeticBuilderAction) => {
+      if (isArithmeticBuilderUpdateResetFocusOverrideAction(action)) {
+        setFocusOverride(null);
+      } else if (isArithmeticBuilderDeleteAction(action)) {
+        const newExpression = deleteToken(expression.text, action);
+        updateExpression?.(newExpression);
+        setExpression(newExpression);
+        if (defined(action.focusOverride)) {
+          setFocusOverride(action.focusOverride);
+        }
+      } else if (isArithmeticBuilderReplaceAction(action)) {
+        const newExpression = replaceToken(expression.text, action);
+        updateExpression?.(newExpression);
+        setExpression(newExpression);
+        if (defined(action.focusOverride)) {
+          setFocusOverride(action.focusOverride);
+        }
```

Comment:
just curious, for my benefit, why you did this refactor?

Reply:
`useReducer` requires that you dispatch an async operation. But in order to sync the internal state of the equations builder and the external state, we ideally update the 2 states simultaneously in 1 render pass (and to avoid having to duplicate the logic twice
- once inside the reducer
- once to propagate the changes out of this component